src/rebuild/scan.c

```diff
@@ -608,9 +612,26 @@ placement_check(uuid_t co_uuid, daos_unit_oid_t oid,
 		 * myrank should always not equal to tgt_rebuild. XXX
 		 */
 		if (myrank != tgts[i]) {
+			daos_unit_oid_t		tmp = oid;
+
+			tmp.id_shard = shards[i];
+			rc = ds_pool_check_leader(rpt->rt_pool_uuid, &tmp,
+						  rpt->rt_rebuild_ver, NULL);
+			if (rc == 0) {
+				/* The leader shard is not on current server,
+				 * then current server does not know whether
+				 * related DTX(s) for current shard have been
+				 * committed or not. Skip the shard that will
+				 * be handled by the leader on another server.
+				 */
+				D_DEBUG(DB_REBUILD, "Skip non-leader replica "
+					DF_UOID".\n", DP_UOID(tmp));
+				continue;
```

Comment:
I thought this is only needed for dtx rebuild, not for regular rebuild?

Reply:
No, it is necessary for object rebuild. With DTX introduced, each server only needs to handle the replica that is the leader. Because the server in-rebuilding will only fetch data from the leader replica to avoid touch the data-record with 'prepared' status which visibility is unknown for non-leader.

Reply:
Hmm, I am not so sure about this. I would rather there is no "leader" during rebuild, which really complicate things, especially if leader fails. Plus we already resolve this in the rebuilding target.   could you please only do it for dtx check?

Reply:
Just discussed with Fan Yong last night, do dtx rebuild happens before the normal rebuild. Does it mean after dtx rebuild, we can fetch the data from any replicas or only from leader?

Reply:
Client can fetch data from any replica as long as related DTX status has been done. But please note that, the leader is per-shard based, not per-object based.

Reply:
I mean after dtx rebuild, there suppose to be no "uncommitted" data over the system? so scan and rebuild can be done from any node?

Reply:
The DTX rebuild on each server is independent from one another. When server-A's DTX rebuild done, the server-B may be not yet. So the subsequent object rebuild logic only can handle the shards that have finished DTX rebuild.

Reply:
I mean after dtx rebuild is done, there suppose no "uncommitted" data, so can we do scan on all replicas? instead of only on leader? Besides this is scan phase, I assume only pulling phase will care if it is the leader? This can save some scan time,  since checking leader seems not cheap here.

Reply:
> I mean after dtx rebuild is done, there suppose no "uncommitted" data

There is no old "uncommitted" data. But there may new ones. During pulling phase, you only get the objects or shards list, not the data record. The object or shard on non-leader may contain 'uncommitted' data record. That will cause the non-leader to return -DER_INPROGRESS to the in-rebuilding record.

Reply:
scan will only scan oid table, even it only scan the object in leader, it can not avoid this "uncommitted" data case during pulling phase.  so why still do leader check here? especially this is not cheap.

Reply:
The leader exactly knows about the DTX visibility, but non-leader does not. When the non-leader hits uncertain DTX, it will ask the RPC sponsor to retry with leader. Since each server resync each own DTX table independently, it is quite possible that the rebuild reading from non-leader will hit uncertain DTX, then retry with leader. Since then, why not just talk with leader directly?

Reply:
But this is pulling phase, so I mean you do not need check this during scan phase.  Btw: since you only have at most 512 dtx records, so this happens very rarely during pulling, so I would rather it will try to pull data from all of replicas during the whole rebuild process, instead only from the leader because of the performance.

Reply:
The issue that: object rebuild is per-shard based, not per-object based. Here, the rebuild iteration returns an object oid, then calls pl_obj_find_rebuild() to parse out the shards to be rebuilt on remote target. But these shards may be not on current server, right? If yes, assume some shard_A resides on server_A (the leader), but the current server cannot guarantee server_A's DTX resync has been done. If we ask the remote target to rebuild shard_A, then it may be blocked by the unfinished DTX resync or get some unexpected result. Our solution is that the server_A will notify the remote target to rebuild shard_A when DTX resync has been done on the server_A.

Reply:
I think all the rebuild I/O will take a timestamp and only enumerate and pull data before the timestamp, which means leader resync should guarantee all updates to be committed right? Also, the leader is not supposed to send the OID to the rebuild target before it completes the force commit, which effectively means when the rebuild target received the OID, it is OK for it to grab data from all replicas?